File: packages/demo-agent/gateway_client.py
import os
import requests
from typing import Optional

class GatewayClient:
    def __init__(self):
        self.base_url = os.environ["GATEWAY_URL"].rstrip("/")
        self.token = os.environ["SESSION_TOKEN"]
        self.session_id = os.environ["SESSION_ID"]

    def _headers(self) -> dict:
        return {"Authorization": f"Bearer {self.token}", "Content-Type": "application/json"}

    def load_messages(self, after_message_id: Optional[str] = None) -> dict:
        payload = {}
        if after_message_id:
            payload["after_message_id"] = after_message_id
        resp = requests.post(f"{self.base_url}/gateway/messages/load", json=payload, headers=self._headers())
        resp.raise_for_status()
        return resp.json()

    def append_messages(self, expected_last_message_id: Optional[str], messages: list) -> dict:
        resp = requests.post(
            f"{self.base_url}/gateway/messages/append",
            json={"expected_last_message_id": expected_last_message_id, "messages": messages},
            headers=self._headers(),
        )
        resp.raise_for_status()
        return resp.json()

    def invoke_llm(self, messages: list, model: str = "glm-5.1") -> dict:
        resp = requests.post(
            f"{self.base_url}/gateway/llm",
            json={"model": model, "messages": messages},
            headers=self._headers(),
        )
        resp.raise_for_status()
        return resp.json()

File: packages/dispatcher/src/gateway-client.ts
export interface AppendResult {
  conversation_id: string
  appended: { id: string; role: string; created_at: string }[]
  last_message_id: string
}

export interface LoadResult {
  conversation_id: string
  messages: { id: string; role: string; content: any[]; source: string; created_at: string }[]
  last_message_id: string | null
}

export function createGatewayClient(gatewayUrl: string, _defaultToken?: string) {
  const base = gatewayUrl.replace(/\/$/, '')

  async function post(path: string, body: unknown, token: string): Promise<any> {
    const res = await fetch(`${base}${path}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(30_000),
    })
    const data = await res.json()
    if (!res.ok) {
      throw new Error(data?.error?.code ?? `HTTP ${res.status}`)
    }
    return data
  }

  return {
    // 把 IM 里收到的用户消息写入会话历史
    async appendMessages(
      expectedLastMessageId: string | null,
      messages: any[],
      token: string
    ): Promise<AppendResult> {
      return post('/gateway/messages/append', { expected_last_message_id: expectedLastMessageId, messages }, token)
    },

    async loadMessages(
      afterMessageId: string | null | undefined,
      token: string
    ): Promise<LoadResult> {
      return post('/gateway/messages/load', afterMessageId ? { after_message_id: afterMessageId } : {}, token)
    },
  }
}
